src/V3EmitCFunc.h

```diff
@@ -0,0 +1,1628 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Emit C++ for tree
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2021 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+
+#ifndef VERILATOR_V3EMITCFUNC_H_
+#define VERILATOR_V3EMITCFUNC_H_
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Global.h"
+#include "V3EmitCBase.h"
+
+#include <algorithm>
+#include <map>
+#include <unordered_set>
+#include <vector>
+
+// Number of VL_CONST_W_*X's in verilated.h (IE VL_CONST_W_8X is last)
+constexpr int EMITC_NUM_CONSTW = 8;
+
+//######################################################################
+// Emit lazy forward declarations
+
+class EmitCLazyDecls final : public AstNVisitor {
+    // NODE STATE/TYPES
+    //  AstNode::user2() -> bool. Already emitted decl for symbols.
+    AstUser2InUse m_inuser2;
+
+    // MEMBERS
+    std::unordered_set<string> m_emittedManually;  // Set of names already declared manually.
+    EmitCBaseVisitor& m_emitter;  // For access to file output
+    bool m_needsBlankLine = false;  // Emit blank line if any declarations were emitted (cosmetic)
+
+    void lazyDeclare(AstCFunc* funcp) {
+        // Already declared in this compilation unit
+        if (funcp->user2SetOnce()) return;
+        // Check if this kind of function is lazily declared
+        if (!(funcp->isMethod() && funcp->isLoose()) && !funcp->dpiImportPrototype()) return;
+        // Already declared manually
+        if (m_emittedManually.count(funcp->nameProtect())) return;
+        // Needs lazy declaration, emit one
+        m_emitter.emitCFuncDecl(funcp, VN_CAST_CONST(funcp->user4p(), NodeModule),
+                                funcp->dpiImportPrototype());
+        m_needsBlankLine = true;
+    }
+
+    void lazyDeclareConstPoolVar(AstVar* varp) {
+        if (varp->user2SetOnce()) return;  // Already declared
+        const string nameProtect
+            = m_emitter.topClassName() + "__ConstPool__" + varp->nameProtect();
+        m_emitter.puts("extern const ");
+        m_emitter.puts(varp->dtypep()->cType(nameProtect, false, false));
+        m_emitter.puts(";\n");
+        m_needsBlankLine = true;
+    }
+
+    virtual void visit(AstNodeCCall* nodep) override {
+        lazyDeclare(nodep->funcp());
+        iterateChildren(nodep);
+    }
+
+    virtual void visit(AstAddrOfCFunc* nodep) override {  //
+        lazyDeclare(nodep->funcp());
+    }
```

Comment:
Even though not really needed:

```suggestion
        lazyDeclare(nodep->funcp());
        iterateChildren(nodep);
    }
```